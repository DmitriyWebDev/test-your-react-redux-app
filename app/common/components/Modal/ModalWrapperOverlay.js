import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

class ModalWrapperOverlay extends React.Component {
  constructor(props) {
    super(props);
    this.handleOverlayClick = this.handleOverlayClick.bind(this);
    this.modalContentRef = React.createRef();
  }

  componentDidMount() {
    document.body.style.overflow = 'hidden';
  }

  componentWillUnmount() {
    document.body.style.overflow = '';
  }

  handleOverlayClick(e) {
    const { shouldCloseOnOverlayClick, onRequestClose } = this.props;

    if (!shouldCloseOnOverlayClick) return null;

    const { target } = e;
    const modalContent = this.modalContentRef.current;
    if (target !== modalContent && target.contains(modalContent)) {
      onRequestClose();
    }
  }

  render() {
    const { modalId, modalClassName, overlayClassName, children } = this.props;

    return (
      <div className={cn('app-modal__window-overlay-wrapper', overlayClassName)} onClick={this.handleOverlayClick}>
        <div id={modalId} className={cn('app-modal__window', modalClassName)} ref={this.modalContentRef}>
          {children}
        </div>
      </div>
    );
  }
}

export default ModalWrapperOverlay;

ModalWrapperOverlay.propTypes = {
  modalId: PropTypes.string,
  modalClassName: PropTypes.string,
  overlayClassName: PropTypes.string,
  onRequestClose: PropTypes.func,
  shouldCloseOnOverlayClick: PropTypes.bool,
  children: PropTypes.node,
};

ModalWrapperOverlay.defaultProps = {
  modalId: '',
  modalClassName: '',
  overlayClassName: '',
  onRequestClose: () => {},
  shouldCloseOnOverlayClick: true,
  children: <span>Modal window default content</span>,
};
