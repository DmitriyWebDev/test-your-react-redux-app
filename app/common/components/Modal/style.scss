$app-modal: '.app-modal';

#{$app-modal} {

  &__window-overlay-wrapper {
    display: block;
    position: fixed;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    overflow: auto;
    z-index: 1030;
    text-align: center;
    background-color: rgba(39, 48, 55, 0.6100000143051147);
  }

  &__window {
    display: block;
    position: relative;
    z-index: 30;
    background-color: #fefefe;
    margin: 30px auto;
    padding: 0;
    box-shadow: 0 16px 26px 0 rgba(0, 0, 0, 0.3100000023841858);
    width: 600px;
  }

  &__window_vertical-centering {
    margin: 0 auto;
    top: 50%;
    transform: translateY(-50%);
  }

}

