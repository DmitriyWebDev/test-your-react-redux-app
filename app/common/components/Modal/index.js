import React from 'react';
import ReactDOM from 'react-dom';
import './style.scss';
import PropTypes from 'prop-types';
import ModalWrapperOverlay from './ModalWrapperOverlay';

class Modal extends React.Component {
  render() {
    if (!this.props.isOpen) return null;

    const { id, className, children, onRequestClose, overlayClassName, shouldCloseOnOverlayClick } = this.props;

    return ReactDOM.createPortal(
      <div className={'app-modal'}>
        <ModalWrapperOverlay
          modalId={id}
          modalClassName={className}
          overlayClassName={overlayClassName}
          onRequestClose={onRequestClose}
          shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
        >
          {children}
        </ModalWrapperOverlay>
      </div>,
      document.body,
    );
  }
}

export default Modal;

Modal.propTypes = {
  isOpen: PropTypes.bool,
  id: PropTypes.string,
  className: PropTypes.string,
  overlayClassName: PropTypes.string,
  onRequestClose: PropTypes.func,
  shouldCloseOnOverlayClick: PropTypes.bool,
  children: PropTypes.node,
};

Modal.defaultProps = {
  isOpen: false,
  id: '',
  className: '',
  overlayClassName: '',
  onRequestClose: () => {},
  shouldCloseOnOverlayClick: true,
};
