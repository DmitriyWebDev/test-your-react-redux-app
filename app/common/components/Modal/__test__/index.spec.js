import React from 'react';
import { shallow } from 'enzyme';
import Modal from '../index';

describe('Modal', () => {
  it('renders correctly, matches the snapshot', () => {
    const ModalComponent = shallow(
      <Modal
        className={'class-1'}
        id={'id-1'}
        isOpen={true}
        onRequestClose={() => ({})}
        overlayClassName={'overlay-class-1'}
        shouldCloseOnOverlayClick={true}
      >
        Modal window content
      </Modal>,
    );
    expect(ModalComponent).toMatchSnapshot();
  });
});
