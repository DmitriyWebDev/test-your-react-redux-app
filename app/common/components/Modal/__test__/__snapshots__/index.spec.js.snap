// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly, matches the snapshot 1`] = `
<Portal
  containerInfo={<body />}
>
  <div
    className="app-modal"
  >
    <ModalWrapperOverlay
      modalClassName="class-1"
      modalId="id-1"
      onRequestClose={[Function]}
      overlayClassName="overlay-class-1"
      shouldCloseOnOverlayClick={true}
    >
      Modal window content
    </ModalWrapperOverlay>
  </div>
</Portal>
`;
